clash-lib/src/Clash/Core/HasType.hs

```diff
@@ -106,6 +108,42 @@ instance InferType Term where
       Cast _ _ a -> a
       Tick _ x -> go x
 
+instance InferType Type where
+  inferCoreTypeOf tcm = go
+   where
+    go (VarTy k) = coreTypeOf k
+    go (ForAllTy _ ty) = go ty
+    go (LitTy NumTy{}) = typeNatKind
+    go (LitTy SymTy{}) = typeSymbolKind
+    go (AnnType _ ty) = go ty
+    go (tyView -> FunTy _ res)
+      | let resTy = go res
+      , isSuperKind tcm resTy = resTy
+
+      | otherwise = liftedTypeKind
+
+    go (tyView -> TyConApp tc args) =
+      foldl' kindFunResult (tyConKind (lookupUniqMap' tcm tc)) args
```

Comment:
```suggestion
      piResultTys tcm (tyConKind (lookupUniqMap' tcm tc)) args
```
Now that it's available. GHC does the same: https://github.com/ghc/ghc/blob/62d720db4f6a53014400a608baf5c56555258eee/compiler/GHC/Core/Type.hs#L2776